File: src/narrative_os/parser.py
"""Parser for .md chapter files with YAML frontmatter."""

from __future__ import annotations
import re
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

import yaml


@dataclass
class DialogueLine:
    character: str
    text: str


@dataclass
class Scene:
    location: str = ""
    timeline: str = ""
    text: str = ""           # full scene text (narrative + dialogue combined)
    dialogue: list[DialogueLine] = field(default_factory=list)
    characters: list[str] = field(default_factory=list)  # characters mentioned in THIS scene
    line_start: int = 0      # line number (0-based) where this scene starts in the body


@dataclass
class Chapter:
    number: int
    title: str
    location: str = ""
    timeline: str = ""
    characters: list[str] = field(default_factory=list)
    pov: str = ""
    status: str = "draft"
    word_target: int = 0
    scenes: list[Scene] = field(default_factory=list)
    raw_text: str = ""
    filename: str = ""


DIALOGUE_RE = re.compile(r'^\[character:\s*([^\]]+)\]\s*[—–-]\s*(.*)', re.IGNORECASE)
DATAVIEW_RE = re.compile(r'^(\w+)::\s*(.+)$')
WIKILINK_RE = re.compile(r'\[\[([^\]]+)\]\]')


def strip_wikilinks(text: str) -> str:
    """[[Name]] -> Name"""
    return WIKILINK_RE.sub(lambda m: m.group(1), text).strip()


def parse_wikilink_list(value: Any) -> list[str]:
    """Parse frontmatter characters/locations list -- strips [[]]."""
    if not value:
        return []
    if isinstance(value, list):
        return [strip_wikilinks(str(v)) for v in value]
    if isinstance(value, str):
        return [strip_wikilinks(value)]
    return []


def parse_chapter(path: Path) -> Chapter | None:
    """Parse a .md chapter file. Returns None on error."""
    try:
        raw = path.read_text(encoding="utf-8")
    except OSError:
        return None

    # Split frontmatter
    fm: dict[str, Any] = {}
    body = raw
    if raw.startswith("---"):
        parts = raw.split("---", 2)
        if len(parts) >= 3:
            try:
                fm = yaml.safe_load(parts[1]) or {}
            except yaml.YAMLError as e:
                fm = {"status": "yaml_error"}
                print(f"Warning: Failed to parse YAML frontmatter in {path}: {e}")
            body = parts[2].lstrip("\n")

    chapter = Chapter(
        number=int(fm.get("chapter", 0)),
        title=str(fm.get("title", path.stem)),
        location=strip_wikilinks(str(fm.get("location", ""))),
        timeline=str(fm.get("timeline", "")),
        characters=parse_wikilink_list(fm.get("characters")),
        pov=strip_wikilinks(str(fm.get("pov", ""))),
        status=str(fm.get("status", "draft")),
        word_target=int(fm.get("word_target", 0)),
        raw_text=body,
        filename=path.name,
    )

    # Parse scenes split by ---
    # Track line numbers by splitting on --- while recording positions
    body_lines = body.splitlines(keepends=True)
    raw_scenes: list[tuple[str, int]] = []  # (raw_scene_text, start_line_0based)
    current_start = 0
    current_block: list[str] = []
    for line_no, line in enumerate(body_lines):
        if re.match(r'^---\s*$', line):
            raw_scenes.append(("".join(current_block), current_start))
            current_block = []
            current_start = line_no + 1
        else:
            current_block.append(line)
    raw_scenes.append(("".join(current_block), current_start))

    current_location = chapter.location
    current_timeline = chapter.timeline

    for raw_scene_text, scene_line_start in raw_scenes:
        raw_scene = raw_scene_text.strip()
        if not raw_scene:
            continue

        scene = Scene(location=current_location, timeline=current_timeline, line_start=scene_line_start)
        lines = raw_scene.splitlines()
        text_lines = []

        # Parse Dataview inline metadata from top of scene block
        i = 0
        while i < len(lines):
            m = DATAVIEW_RE.match(lines[i].strip())
            if m:
                key, val = m.group(1).lower(), m.group(2).strip()
                if key == "location":
                    scene.location = strip_wikilinks(val)
                    current_location = scene.location
                elif key == "timeline":
                    scene.timeline = strip_wikilinks(val)
                    current_timeline = scene.timeline
                i += 1
            else:
                break

        # Parse remaining lines
        for line in lines[i:]:
            dm = DIALOGUE_RE.match(line.strip())
            if dm:
                char = dm.group(1).strip()
                text = dm.group(2).strip()
                scene.dialogue.append(DialogueLine(character=char, text=text))
                if char not in scene.characters:
                    scene.characters.append(char)
                text_lines.append(f"[{char}]: {text}")
            else:
                text_lines.append(line)

        scene.text = "\n".join(text_lines).strip()
        if scene.text:
            chapter.scenes.append(scene)

    return chapter


def parse_book(book_dir: Path) -> list[Chapter]:
    """Parse all .md chapter files in book_dir/chapters/ (or book_dir itself)."""
    chapters_dir = book_dir / "chapters"
    if chapters_dir.exists():
        search_dir = chapters_dir
    else:
        search_dir = book_dir

    chapters = []
    for path in sorted(search_dir.glob("*.md")):
        if path.name.startswith("_"):
            continue
        ch = parse_chapter(path)
        if ch and ch.scenes:
            chapters.append(ch)

    chapters.sort(key=lambda c: c.number)
    return chapters

File: obsidian-plugin/src/parser.ts
import { parseYaml } from "obsidian";

export interface DialogueLine {
  character: string;
  text: string;
}

export interface Scene {
  location: string;
  timeline: string;
  text: string;
  dialogue: DialogueLine[];
  characters: string[];
  line_start: number;
}

export interface Chapter {
  number: number;
  title: string;
  location: string;
  timeline: string;
  characters: string[];
  pov: string;
  status: string;
  word_target: number;
  scenes: Scene[];
  raw_text: string;
  filename: string;
}

const DIALOGUE_RE = /^\[character:\s*([^\]]+)\]\s*[—–-]\s*(.*)/i;
const DATAVIEW_RE = /^(\w+)::\s*(.+)$/;
const WIKILINK_RE = /\[\[([^\]]+)\]\]/g;

export function stripWikilinks(text: string): string {
  return text.replace(WIKILINK_RE, "$1").trim();
}

export function parseWikilinkList(value: unknown): string[] {
  if (!value) return [];
  if (Array.isArray(value)) {
    return value.map((v) => stripWikilinks(String(v)));
  }
  if (typeof value === "string") {
    return [stripWikilinks(value)];
  }
  return [];
}

const FM_RE = /^---\r?\n([\s\S]*?)\n---\r?\n?([\s\S]*)$/;

export function parseChapter(rawText: string, filename: string): Chapter | null {
  let fm: Record<string, any> = {};
  let body = rawText;

  const fmMatch = rawText.match(FM_RE);
  if (fmMatch) {
    try {
      fm = parseYaml(fmMatch[1]) || {};
    } catch (e) {
      fm = { status: "yaml_error" };
      console.warn(`Failed to parse YAML frontmatter in ${filename}:`, e);
    }
    body = fmMatch[2].replace(/^\n/, "");
  }

  const stem = filename.replace(/\.md$/, "");

  const chapter: Chapter = {
    number: Number(fm.chapter) || 0,
    title: String(fm.title || stem),
    location: stripWikilinks(String(fm.location || "")),
    timeline: String(fm.timeline || ""),
    characters: parseWikilinkList(fm.characters),
    pov: stripWikilinks(String(fm.pov || "")),
    status: String(fm.status || "draft"),
    word_target: Number(fm.word_target) || 0,
    scenes: [],
    raw_text: body,
    filename: filename,
  };

  const bodyLines = body.split("\n");
  const rawScenes: Array<{ text: string; start: number }> = [];
  let currentStart = 0;
  let currentBlock: string[] = [];

  for (let lineNo = 0; lineNo < bodyLines.length; lineNo++) {
    const line = bodyLines[lineNo];
    if (/^---\s*$/.test(line)) {
      rawScenes.push({ text: currentBlock.join("\n"), start: currentStart });
      currentBlock = [];
      currentStart = lineNo + 1;
    } else {
      currentBlock.push(line);
    }
  }
  rawScenes.push({ text: currentBlock.join("\n"), start: currentStart });

  let currentLocation = chapter.location;
  let currentTimeline = chapter.timeline;

  for (const rawScene of rawScenes) {
    const rawSceneText = rawScene.text.trim();
    if (!rawSceneText) continue;

    const scene: Scene = {
      location: currentLocation,
      timeline: currentTimeline,
      line_start: rawScene.start,
      text: "",
      dialogue: [],
      characters: [],
    };

    const lines = rawScene.text.split("\n");

    // Extract dataview metadata
    for (let i = 0; i < lines.length; i++) {
      const trimmed = lines[i].trim();
      if (!trimmed) continue; // skip blank lines before/between metadata
      const match = trimmed.match(DATAVIEW_RE);
      if (match) {
        const key = match[1].toLowerCase();
        const val = match[2].trim();
        if (key === "location") {
          scene.location = stripWikilinks(val);
          currentLocation = scene.location;
        } else if (key === "timeline") {
          scene.timeline = stripWikilinks(val);
          currentTimeline = scene.timeline;
        }
      } else {
        break; // first non-blank non-Dataview line ends metadata block
      }
    }

    // Extract dialogue and characters
    for (let j = 0; j < lines.length; j++) {
      const line = lines[j];
      const dm = line.trim().match(DIALOGUE_RE);
      if (dm) {
        const char = dm[1].trim();
        const text = dm[2].trim();
        scene.dialogue.push({ character: char, text });
        if (!scene.characters.includes(char)) {
          scene.characters.push(char);
        }
      }
    }

    // Crucially: save the EXACT raw text so edit_scene string matching works
    scene.text = rawScene.text.trim();
    if (scene.text) {
      chapter.scenes.push(scene);
    }
  }

  return chapter;
}
